backend/test_observer/controllers/test_executions/end_test.py

```diff
@@ -0,0 +1,181 @@
+from fastapi import APIRouter, Depends, HTTPException
+from sqlalchemy import select
+from sqlalchemy.orm import Session, joinedload
+
+from test_observer.data_access.models import (
+    Artefact,
+    ArtefactBuild,
+    Stage,
+    TestCase,
+    TestExecution,
+    TestResult,
+)
+from test_observer.data_access.models_enums import (
+    FamilyName,
+    TestExecutionReviewDecision,
+    TestExecutionStatus,
+    TestResultStatus,
+)
+from test_observer.data_access.repository import get_or_create
+from test_observer.data_access.setup import get_db
+
+from .logic import delete_previous_results
+from .models import C3TestResult, C3TestResultStatus, EndTestExecutionRequest
+
+router = APIRouter()
+
+
+@router.put("/end-test")
+def end_test_execution(request: EndTestExecutionRequest, db: Session = Depends(get_db)):
+    test_execution = _find_related_test_execution(request, db)
+
+    if test_execution is None:
+        raise HTTPException(status_code=404, detail="Related TestExecution not found")
+
+    delete_previous_results(db, test_execution)
+    _store_test_results(db, request.test_results, test_execution)
+
+    has_failures = test_execution.has_failures
+
+    test_execution.status = (
+        TestExecutionStatus.FAILED if has_failures else TestExecutionStatus.PASSED
+    )
+
+    prev_test_execution = _get_previous_test_execution(db, test_execution)
+
+    if (
+        prev_test_execution
+        and not has_failures
+        and prev_test_execution.is_approved
+        and _ran_all_previously_run_cases(prev_test_execution, test_execution)
```

Comment:
On one of the previous standups we mentioned making sure that the set of SKIPPED tests hasn't changed, to help us make sure not a lot of test cases started getting skipped instead of passing. Should we add that verification here as well?

Reply:
This is partially covered by `_ran_all_previously_run_cases`. In the sense that any tests that have run before (whether passed or failed) have to have run this time too. Together with `not has_failures` ensures that any test that ran before must have passed this time and not skipped. 

But it could be that there are new tests (didn't run last time) that have been skipped. This sounds fine to me, but if you are worried about this, then we can discuss it with others.

Reply:
Ah I didn't notice the condition`if tr.status != TestResultStatus.SKIPPED` which actually accounts for what I was mentioning. Sorry for this.

In terms of new test cases, I'm not sure what would be the best way to resolve them. In some cases we might want to check explicitly the outcome of the new test cases. I wonder if we shouldn't approve executions where there are any additional test cases compared to previous run, or this wouldn't need any attention of the reviewer?

Reply:
At least no one brought this up in the spec. I assume it's safe to have new test cases that pass or skip. But we can definitely add this check if we decided it's important later on.